scripts/test.sh

```diff
@@ -17,85 +17,60 @@
 #See the License for the specific language governing permissions and
 #limitations under the License.
 
-# The script does automatic checking on a Go package and its sub-packages, including:
-# 1. gofmt         (http://golang.org/cmd/gofmt/)
-# 2. goimports     (https://github.com/bradfitz/goimports)
-# 3. golint        (https://github.com/golang/lint) (disabled)
-# 4. go vet        (http://golang.org/cmd/vet) (disabled)
-# 5. race detector (http://blog.golang.org/race-detector) (disabled)
-# 6. test coverage (http://blog.golang.org/cover)
-
-if [[ $# -ne 1 ]]; then
-	echo "ERROR; missing SNAP_TEST_TYPE (Usage: $0 SNAP_TEST_TYPE)"
-	exit -2
-elif [[ “$1” != “legacy” && "$1" != "small" && “$1” != “medium” && “$1” != “large” ]]; then
-	echo "Error; invalid SNAP_TEST_TYPE (value must be one of 'legacy', 'small', 'medium', or 'large'; received $1)"
-	exit -1
-fi
-SNAP_TEST_TYPE=$1
+# Support travis.ci environment matrix:
+SNAP_TEST_TYPE="${SNAP_TEST_TYPE:-$1}"
 
-# If the following plugins don't exist, exit
-[ -f $SNAP_PATH/plugin/snap-collector-mock1 ] || { echo 'Error: $SNAP_PATH/plugin/snap-collector-mock1 does not exist. Run make to build it.' ; exit 1; }
-[ -f $SNAP_PATH/plugin/snap-collector-mock2 ] || { echo 'Error: $SNAP_PATH/plugin/snap-collector-mock2 does not exist. Run make to build it.' ; exit 1; }
-[ -f $SNAP_PATH/plugin/snap-processor-passthru ] || { echo 'Error: $SNAP_PATH/plugin/snap-processor-passthru does not exist. Run make to build it.' ; exit 1; }
-[ -f $SNAP_PATH/plugin/snap-publisher-file ] || { echo 'Error: $SNAP_PATH/plugin/snap-publisher-file does not exist. Run make to build it.' ; exit 1; }
-
-TEST_DIRS="cmd/ control/ core/ mgmt/ pkg/ snapd.go scheduler/"
-# VET_DIRS="./cmd/... ./control/... ./core/... ./mgmt/... ./pkg/... ./scheduler/... ."
+UNIT_TEST="${UNIT_TEST:-"gofmt goimports go_test go_cover"}"
 
 set -e
+set -u
+set -o pipefail
+
+__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+__proj_dir="$(dirname "$__dir")"
+
+SNAP_PATH="${SNAP_PATH:-"${__proj_dir}/build"}"
+export SNAP_PATH
+test_dirs=$(find . -type f -name '*.go' -not -path "./.*" -not -path "*/_*" -not -path "./Godeps/*" -not -path "./vendor/*" -print0 | xargs -0 -n1 dirname| sort -u)
+export test_dirs
+
+# shellcheck source=scripts/common.sh
+. "${__dir}/common.sh"
+
+_debug "script directory ${__dir}"
+_debug "project directory ${__proj_dir}"
+
+[[ "$SNAP_TEST_TYPE" =~ ^(small|medium|large|legacy)$ ]] || _error "invalid TEST_TYPE (value must be 'small', 'medium', 'large', or 'legacy', recieved:${SNAP_TEST_TYPE}"
```

Comment:
Probably should say "Invalid SNAP_TEST_TYPE", though should a different error be thrown if SNAP_TEST_TYPE is not set?

And at the end of the line, recieved -> received.